Assets/MRTK/Services/CameraSystem/ProjectionOverride.cs

```diff
@@ -0,0 +1,121 @@
+        /// </summary>
+        [StructLayout(LayoutKind.Sequential)]
+        private struct HolographicFrameNativeData
+        {
+#pragma warning disable 0649
+            public uint VersionNumber;
+            public uint MaxNumberOfCameras;
+            public IntPtr ISpatialCoordinateSystemPtr;
+            public IntPtr IHolographicFramePtr;
+            public IntPtr IHolographicCameraPtr;
+#pragma warning restore 0649
+        }
+
+        /// <summary>
+        /// Coroutine function to set the camera matrices back to their defaults
+        /// </summary>
+        /// <returns>Coroutine enumerator</returns>
+        IEnumerator ResetViewMatricesOnFrameEnd()
+        {
+            yield return new WaitForEndOfFrame();
+            Camera cam = GetComponent<Camera>();
+            cam.ResetStereoViewMatrices();
+            cam.ResetStereoProjectionMatrices();
+        }
+
+        /// <inheritdoc />
+        void OnPreCull()
+        {
+            if (!ReadingModeEnabled)
+            {
+                return;
+            }
+
+            StartCoroutine(ResetViewMatricesOnFrameEnd());
+
+            IntPtr nativeStruct = UnityEngine.XR.XRDevice.GetNativePtr();
+            if (nativeStruct != IntPtr.Zero)
+            {
+                HolographicFrameNativeData s = System.Runtime.InteropServices.Marshal.PtrToStructure<HolographicFrameNativeData>(nativeStruct);
+
+                if (s.IHolographicFramePtr != IntPtr.Zero)
+                {
+                    var holographicCamera = (HolographicCamera)System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(s.IHolographicCameraPtr);
+                    var holographicFrame = (HolographicFrame)System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(s.IHolographicFramePtr);
+                    HolographicFramePrediction prediction = holographicFrame.CurrentPrediction;
+
+                    if (holographicCamera.CanOverrideViewport)
+                    {
+                        for (int i = 0; i < prediction.CameraPoses.Count; ++i)
+                        {
+                            HolographicCameraPose cameraPose = prediction.CameraPoses[i];
+
+                            // Default spatial locator is the HMD
+                            SpatialLocator defaultSpatialLocator = SpatialLocator.GetDefault();
+
+                            SpatialLocatorAttachedFrameOfReference attachedFrameOfReference = defaultSpatialLocator.CreateAttachedFrameOfReferenceAtCurrentHeading();
+                            SpatialCoordinateSystem attachedCoordinateFrame = attachedFrameOfReference.GetStationaryCoordinateSystemAtTimestamp(prediction.Timestamp);
+
+                            HolographicStereoTransform stereoProjection = cameraPose.ProjectionTransform;
+                            float ResolutionScale = 45.0f / 33.0f;
```

Comment:
What does this represent? Can we cache it as a `const`? Is it something we'd want configurable?

Reply:
HoloLens 2's default SoC render resolution is approximately 33 pixels per degree (ppd), averaged over the field of view.  The actual display generates 45 laser pulses per degree (averaged over the field of view).  This constant is specifying the ratio you need to scale the focal length of the projection matrix by to achieve a field of view which will cause the SoC to produce a 45ppd image.  There is no particular reason why this should be an upper limit.  It is likely the sweet spot where increasing the SoC render FoV has diminishing returns on text quality.  I would expect image quality to improve all the way up to at least 66ppd - although your FoV would be tiny in that case.

Reply:
Maybe allow the user to specify the PPD they want to render at, with comments stating that 45ppd is the approximate native resolution of the display and likely the sweet spot between FoV and legibility.

Reply:
Hey @wiwei!
About 

> There is currently a bug (as of Unity 2019.4.5) where Unity doesn't fully use the adjusted projection when the "Stereo Rendering Mode" is "Single Pass Instanced."